### PR TITLE
PP-4374 Introduced SmartpayNotificationService

### PR DESCRIPTION
**Pull request by kbottla**
## WHAT

Part of refactoring notification handling (see https://github.com/alphagov/pay-connector/pull/787 for the original full attempt...). Introducing gateway specific service objects to handle notifications instead of generalised notification handler (`NotificationService`). We don't need a generalised `NotificationService` as each gateway notification has dedicated notification resource endpoint. Having separate service object will make it easier to understand and follow through the path.  

- Added `SmartpayNotificationService` to delegate Smartpay notification processing to service instead of a generic handler.
- NotificationResource for Smartpay notifications now uses `SmartNotificationService`
- Also removed handleNotification() from NotificationResource as it is no longer required

Note that any existing code that handles Smartpay notifications still stays and will be removed in subsequent PRs

## HOW 
- Existing tests in `SmartpayNotificationResourceWithAccountSpecificAuthITest` should continue to pass
- All tests in `SmartpayNotificationServiceTest` should pass

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/gateway/smartpay/SmartpayNotificationService.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+    private final ObjectMapper objectMapper;
+    private final ChargeDao chargeDao;
+    private final ChargeNotificationProcessor chargeNotificationProcessor;
+    private final RefundNotificationProcessor refundNotificationProcessor;
+
+    private static final String PAYMENT_GATEWAY_NAME = SMARTPAY.getName();
+
+    @Inject
+    public SmartpayNotificationService(ObjectMapper objectMapper, ChargeDao chargeDao,
```
Comment (georgeracu) [inline src/main/java/uk/gov/pay/connector/gateway/smartpay/SmartpayNotificationService.java]: `ObjectMapper` can be removed from the constructor and can be created when needed. It looks like is being used only in 1 place in this class.

File: src/main/java/uk/gov/pay/connector/webhook/resource/NotificationResource.java [hunk 2]
```diff
@@ -45,7 +43,10 @@ public NotificationResource(WorldpayNotificationService worldpayNotificationServ
     @PermitAll
     @Path("/v1/api/notifications/smartpay")
     public Response authoriseSmartpayNotifications(String notification) {
-        return handleNotification("not-required", "smartpay", notification);
+        smartpayNotificationService.handleNotificationFor(notification);
+        String response = "[accepted]";
+        logger.info("Responding to notification from provider={} with 200 {}", "smartpay", response);
```
Comment (georgeracu) [inline src/main/java/uk/gov/pay/connector/webhook/resource/NotificationResource.java]: Inlining `smartpay` would make it simpler

File: src/test/java/uk/gov/pay/connector/util/TestTemplateResourceLoader.java [hunk 3]
```diff
@@ -69,11 +69,11 @@
     public static final String SMARTPAY_VALID_CAPTURE_SMARTPAY_REQUEST = SMARTPAY_BASE_NAME + "/valid-capture-smartpay-request.xml";
 
     public static final String SMARTPAY_MULTIPLE_NOTIFICATIONS = SMARTPAY_BASE_NAME + "/multiple-notifications.json";
-    public static final String SMARTPAY_MULTIPLE_NOTIFICATIONS_DIFFERENT_DATES = SMARTPAY_BASE_NAME + "/multiple-notifications-different-dates.xml";
-    public static final String SMARTPAY_NOTIFICATION_AUTHORISATION = SMARTPAY_BASE_NAME + "/notification-authorisation.xml";
-    public static final String SMARTPAY_NOTIFICATION_CAPTURE_WITH_UNKNOWN_STATUS = SMARTPAY_BASE_NAME + "/notification-capture-with-unknown-status.xml";
-    public static final String SMARTPAY_NOTIFICATION_CAPTURE = SMARTPAY_BASE_NAME + "/notification-capture.xml";
-    public static final String SMARTPAY_NOTIFICATION_REFUND = SMARTPAY_BASE_NAME + "/notification-refund.xml";
+    public static final String SMARTPAY_MULTIPLE_NOTIFICATIONS_DIFFERENT_DATES = SMARTPAY_BASE_NAME + "/multiple-notifications-different-dates.json";
```
Comment (georgeracu) [inline src/test/java/uk/gov/pay/connector/util/TestTemplateResourceLoader.java:72]: I wonder how were this working with `.xml` as I see in master they are `.json` 🤔

**Review by georgeracu — CHANGES_REQUESTED** [review]
Nothing major, just a comment on removing a constructor argument that is not required to be injected. Nice one 👍

File: src/main/java/uk/gov/pay/connector/gateway/smartpay/SmartpayNotificationService.java (hunk 1 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/gateway/smartpay/SmartpayNotificationService.java]: 👍

File: src/test/java/uk/gov/pay/connector/util/TestTemplateResourceLoader.java (hunk 3 above)
Comment (kbottla) [reply src/test/java/uk/gov/pay/connector/util/TestTemplateResourceLoader.java:72]: These constants were never used and were hardcoded 
https://github.com/alphagov/pay-connector/blob/master/src/test/java/uk/gov/pay/connector/it/resources/smartpay/SmartpayNotificationResourceWithAccountSpecificAuthITest.java#L227

File: src/main/java/uk/gov/pay/connector/webhook/resource/NotificationResource.java (hunk 2 above)
Comment (kbottla) [reply src/main/java/uk/gov/pay/connector/webhook/resource/NotificationResource.java]: 👍

**Review by georgeracu — APPROVED** [review]
LGTM

**Pull request merged**